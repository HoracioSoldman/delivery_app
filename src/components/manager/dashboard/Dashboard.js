import React, {Component} from 'react'
import {connect} from 'react-redux'
import PropTypes from 'prop-types'
import {fetchShipments, updateAllShipments} from '../../../redux/actions/shipment-actions'
import {fetchBikers, updateAllBikers, isLoggedIn} from '../../../redux/actions/user-actions'
import Overview from '../../list-overview/Overview'
import Header from '../../Header'
import {PageTitle} from '../../stylus'
import Table from '../../list-table/Table'

class Dashboard extends Component {

    componentDidMount(){
        this.props.fetchShipments();
        this.props.fetchBikers();
    }

    componentWillUnmount(){
        this.props.updateAllShipments([]);
        this.props.updateAllBikers([]);
    }
    
    render(){
        const {user} = isLoggedIn();
        const summary = [
            {title: 'SHIPMENTS', nbr:this.props.shipments.length, primaryColor:'#607D8B', secondaryColor: '#78909C'},
            {title: 'WAITING', nbr:0, primaryColor:'#673AB7', secondaryColor: '#7E57C2'},
            {title: 'ASSIGNED', nbr:0, primaryColor:'#2196F3', secondaryColor: '#42A5F5'},
            {title: 'PICKED_UP', nbr:0, primaryColor:'#009688', secondaryColor: '#26A69A'},
            {title: 'DELIVERED', nbr:0, primaryColor:'#4CAF50', secondaryColor: '#66BB6A'},
        ];
    

        this.props.shipments.forEach((shp, i)=>{
            switch(shp.status){
                case 'WAITING':
                    summary[1].nbr++;
                break;
                
                case 'ASSIGNED':
                    summary[2].nbr++;
                break;
                
                case 'PICKED_UP':
                    summary[3].nbr++;
                break;
                
                case 'DELIVERED':
                    summary[4].nbr++;
                break;
    
                default:
                    summary[0].nbr = this.props.shipments.length; 
                    break;
    
            }
            
        })

        const table_header = ['#', 'ID', 'Origin', 'Destination', 'Status', 'Assignee', 'Pick Up', 'Delivery'];
        //Filter the list according to what has been clicked on the Overview list
        //It Can be SHIPMENTS, WAITING,...
        const table_body = this.props.filter === '' || this.props.filter === 'SHIPMENTS' ?
                    this.props.shipments:
                    this.props.shipments.filter(shp =>(shp.status === this.props.filter));
    
        return (
            <>
                <Header />
                <PageTitle>Shipments Dashboard</PageTitle>
                
                <Overview items={summary}/>
                
                <Table body={table_body} header={table_header} user_role={user.role}/>
                
            </>
        )
    }
}

Dashboard.propTypes = {
    fetchShipments: PropTypes.func.isRequired,
    fetchBikers: PropTypes.func.isRequired,
    updateAllShipments: PropTypes.func.isRequired,
    updateAllBikers: PropTypes.func.isRequired,
    shipments: PropTypes.array.isRequired,
    filter: PropTypes.string
}

const mapStateToProps = state => ({
    shipments: state.shipments,
    filter: state.filter
})

export default connect(mapStateToProps, 
    {fetchShipments, fetchBikers, updateAllShipments, updateAllBikers}
    )(Dashboard);

export {Dashboard};