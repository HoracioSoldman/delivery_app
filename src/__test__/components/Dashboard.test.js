import React from 'react'
import {shallow} from 'enzyme'
import toJson from 'enzyme-to-json'
import {Dashboard} from '../../components/manager/dashboard/Dashboard'


describe('Dashboard', () => {
    it('renders without crashing', ()=> {
        const props = {
            fetchShipments: jest.fn(),
            fetchBikers: jest.fn(),
            updateAllShipments: jest.fn(),
            updateAllBikers: jest.fn(),
            shipments: [],
            filter: ''
            
        }
        const wrapper = shallow(<Dashboard  {...props} />)
        expect(toJson(wrapper)).toMatchSnapshot() 
    })
    
})
