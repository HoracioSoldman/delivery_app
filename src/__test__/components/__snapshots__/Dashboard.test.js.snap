// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders without crashing 1`] = `
<Fragment>
  <Connect(withRouter(Header)) />
  <styled.h2>
    Shipments Dashboard
  </styled.h2>
  <Overview
    items={
      Array [
        Object {
          "nbr": 0,
          "primaryColor": "#607D8B",
          "secondaryColor": "#78909C",
          "title": "SHIPMENTS",
        },
        Object {
          "nbr": 0,
          "primaryColor": "#673AB7",
          "secondaryColor": "#7E57C2",
          "title": "WAITING",
        },
        Object {
          "nbr": 0,
          "primaryColor": "#2196F3",
          "secondaryColor": "#42A5F5",
          "title": "ASSIGNED",
        },
        Object {
          "nbr": 0,
          "primaryColor": "#009688",
          "secondaryColor": "#26A69A",
          "title": "PICKED_UP",
        },
        Object {
          "nbr": 0,
          "primaryColor": "#4CAF50",
          "secondaryColor": "#66BB6A",
          "title": "DELIVERED",
        },
      ]
    }
  />
  <Table
    body={Array []}
    header={
      Array [
        "#",
        "ID",
        "Origin",
        "Destination",
        "Status",
        "Assignee",
        "Pick Up",
        "Delivery",
      ]
    }
  />
</Fragment>
`;
